### PR TITLE
[Emu3] Emu3 model inference and SFT code, rebased master

### PR DESCRIPTION
**Pull request by chenyingshu**
# What does this PR do?

master-rebased version of PR #913

Adds # (feature) Emu3 model in `examples/emu3`
- inference script for (1) image reconstruction, (2) text-to-image inference, (3) VQA inference
- sft code and scripts for (1) text-to-image and (2) vqa/qa tasks.
- Differences from original [Emu3](https://github.com/baaivision/Emu3):
    - apply sequence parallel training to mindspore version.
    - mindspore ver. supports vqa/text-qa SFT, Emu3 repo does not provide code for vqa sft.

Adds # (feature) in `mindone/transformers` add:
- new logit processor: UnbatchedClassifierFreeGuidanceLogitsProcessor in `mindone/transformers/generation/logits_process.py` 
- new scheduler: get_cosine_with_min_lr_schedule_with_warmup in `mindone/transformers/optimization.py`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/mindspore-lab/mindone/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the documentation with your changes? E.g. record bug fixes or new features in `What's New`. Here are the
      [documentation guidelines](https://github.com/mindspore-lab/mindcv/wiki/%E6%96%87%E6%A1%A3%E7%BC%96%E5%86%99%E8%A7%84%E8%8C%83)
- [x] Did you build and run the code without any errors?
- [x] Did you report the running environment (NPU type/MS version) and performance in the doc? (better record it for data loading, model inference, or training tasks)
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@SamitHuang @CaitinZhao

### CODE REVIEW

**Review by CaitinZhao — APPROVED** [review]

File: examples/emu3/README.md [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+|mindspore |	ascend driver | firmware | cann tookit/kernel|
+|--- | --- | --- | --- |
+|2.5.0 | 24.1RC2 | 7.3.0.1.231 | 8.0.RC3.beta1|
+
+### Dependencies
+
+Enter this directory and install required packages:
+
+```shell
+cd examples/emu3
```
Comment (CaitinZhao) [inline examples/emu3/README.md:53]:
```suggestion
cd mindone
pip install -e .[training]
cd examples/emu3
pip install -r requirements.txt
```
and delete the sys.path.insert in python file

File: examples/emu3/autoencode.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+
+import os
+import os.path as osp
+import sys
+import time
+
+# debug use, TODO: delete later
```
Comment (CaitinZhao) [inline examples/emu3/autoencode.py:8]: delete it

File: examples/emu3/autoencode.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+# VIDEO_FRAMES_PATH = "assets/video_pd"
+
+video = os.listdir(VIDEO_FRAMES_PATH)
+video.sort()
+video = [Image.open(osp.join(VIDEO_FRAMES_PATH, v)) for v in video]
+
+images = processor(video, return_tensors="np")["pixel_values"]
+images = Tensor(images).unsqueeze(0)  # [1, Frames, C, H, W]
+
+# image autoencode #
```
Comment (CaitinZhao) [inline examples/emu3/autoencode.py:72]: delete?

File: examples/emu3/emu3/mllm/processing_emu3.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+import mindspore as ms
+from mindspore import Tensor, _no_grad, jit_class, mint, ops
+
+from .utils_emu3 import Emu3PrefixConstrainedLogitsHelper
+
+logger = logging.get_logger(__name__)
+
+
+@jit_class
+class no_grad(_no_grad):
```
Comment (CaitinZhao) [inline examples/emu3/emu3/mllm/processing_emu3.py:37]: from mindone.diffusers.training_utils import pynative_no_grad

File: examples/emu3/emu3/tests/test_emu3_infer.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import time
+
+from emu3.mllm import Emu3ForCausalLM
+from emu3.mllm.configuration_emu3 import Emu3Config
+from PIL import Image
+
+import mindspore as ms
+from mindspore import nn
+
+ms.set_context(mode=ms.PYNATIVE_MODE, pynative_synchronize=True)
```
Comment (CaitinZhao) [inline examples/emu3/emu3/tests/test_emu3_infer.py:13]:
```suggestion
ms.set_context(mode=ms.PYNATIVE_MODE)
```

File: examples/emu3/emu3/train/prepare_vision_data.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+from emu3.tokenizer import Emu3VisionVQImageProcessor, Emu3VisionVQModel
+from PIL import Image
+from tqdm import tqdm
+
+import mindspore as ms
+from mindspore import _no_grad, jit_class
+
+
+@jit_class
+class no_grad(_no_grad):
```
Comment (CaitinZhao) [inline examples/emu3/emu3/train/prepare_vision_data.py:49]: ditto

File: examples/emu3/emu3/train/prepare_vqa_data.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+from emu3.tokenizer import Emu3VisionVQImageProcessor, Emu3VisionVQModel
+from PIL import Image
+from tqdm import tqdm
+
+import mindspore as ms
+from mindspore import _no_grad, jit_class
+
+
+@jit_class
+class no_grad(_no_grad):
```
Comment (CaitinZhao) [inline examples/emu3/emu3/train/prepare_vqa_data.py:67]: ditto

File: examples/emu3/README.md (hunk 1 above)
Comment (chenyingshu) [reply examples/emu3/README.md:55]: edited.

File: examples/emu3/autoencode.py (hunk 2 above)
Comment (chenyingshu) [reply examples/emu3/autoencode.py]: deleted

File: examples/emu3/autoencode.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+from mindone.utils.amp import auto_mixed_precision
+
+# Both modes are supported
+ms.set_context(mode=ms.PYNATIVE_MODE)  # PYNATIVE
+# ms.set_context(mode=ms.GRAPH_MODE)      # GRAPH
+
+
+@jit_class
+class no_grad(_no_grad):
```
[the remark this answers is not in the document]
Comment (chenyingshu) [reply examples/emu3/autoencode.py]: edited

File: examples/emu3/emu3/mllm/processing_emu3.py (hunk 4 above)
Comment (chenyingshu) [reply examples/emu3/emu3/mllm/processing_emu3.py]: edited

File: examples/emu3/emu3/train/prepare_vision_data.py (hunk 6 above)
Comment (chenyingshu) [reply examples/emu3/emu3/train/prepare_vision_data.py]: edited

File: examples/emu3/emu3/train/prepare_vqa_data.py (hunk 7 above)
Comment (chenyingshu) [reply examples/emu3/emu3/train/prepare_vqa_data.py]: edited

**Review by CaitinZhao — APPROVED** [review]

**Review by SamitHuang — APPROVED** [review]

File: examples/emu3/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+transformers>=4.44.0
+tiktoken==0.6.0
+pillow
+omegaconf
+torch
+opencv-python
+ezcolorlog
+mindcv
```
Comment (SamitHuang) [inline examples/emu3/requirements.txt:8]: state mindcv version
